### ISSUE TITLE
Avoid synchronous operations implementation

### ISSUE THREAD

**Issue by amazuerar** [markusfisch/BinaryEye#210]
Dear Developer!

My name is Alejandro Mazuera-Rozo, I am a PhD Student at Universidad de los Andes, and at Università della Svizzera italiana. I am part of a research on the usage of network libraries within Android apps. As result of this we identified some code locations that might have network related problems.

In this case, we address the code locations that are related to a potential issue because of implementing synchronous operations. As you might know, synchronous requests are not recommended and its usage is discouraged.

1. In **synchronous operations**, tasks are performed one at a time and only when one is completed, the following is unblocked.

2. In **asynchronous operations**, you can move to another task before the previous one finishes, allowing you to deal with multiple requests simultaneously.
           
Moreover, by avoiding this approach, it is less likely to perform network operations on the main thread. Android requires you to perform network operations on a thread other than the main UI thread; a `NetworkOnMainThreadException` is thrown otherwise.
       
In order to address this issue we recommend you to:

1. See: https://developer.android.com/training/basics/network-ops/connecting
2. Consider alternative libraries (e.g., Retrofit) handling asynchronous operations

**Potential Code Location with synchronous operations:**

https://github.com/markusfisch/BinaryEye/blob/af70d9507ba19739064949dd089de9b7112bf255/app/src/main/kotlin/de/markusfisch/android/binaryeye/net/ScanSender.kt#L89

**Comment by markusfisch** [comment]
Hi Alejandro,

Thanks for your interest in this app and for pointing out why it's important to do certain things asynchronous. Really appreciated!

It's just that this specific find doesn't run on the main thread 😉 It's not immediate clear and should probably be refactored to make this clear, but `request()` is a `private` function (meaning that it can only be called within this file) that is called by `Scan.send()` which is also private, which in turn only gets called by `Scan.sendAsync()`. The very first function in that file and the only one that's not private (to this file).

And obviously, `Scan.sendAsync()` doesn't run on the main thread as the "Async" suffix implies 😉 

So the request runs on `Dispatcher.IO` as you can see by looking at `Scan.sendAsync()`.

Also, I want to point out that making something run on a different thread isn't _always_ required nor desirebale. Quite often, you need the result from one operation for the next. In these case, it doesn't make too much sense to use more than one extra thread (keeping the main thread for handling UI of course). I know it has become very easy to spawn new threads on demand, but parallelism is always a complexity, even when it's hidden behind very minimal syntax. So it's best to limit spawing threads to whenever there's a reason for it - and not by default for anything.
